src/mem/metaslab.h

```diff
@@ -272,7 +272,7 @@ namespace snmalloc
      */
     SNMALLOC_FAST_PATH
     MetaEntry(Metaslab* meta, uintptr_t remote_and_sizeclass)
-    : meta(reinterpret_cast<uintptr_t>(meta)),
+    : meta(unsafe_to_uintptr<Metaslab>(meta)),
```

Comment:
Would it be better to use a union of the allowed types here, rather than an implicit union in the form of `uintptr_t`?

Reply:
It needs to be `uintptr_t` as the borrowing bottom bits and allowing nullptr leads to UB if it is real pointer.
For instance, a redblack tree node with a null right child would store 258 (I think), and adding 258 to nullptr is UB.